File: Project 1/Employee.cs
// Garland Lau
// CPSC 3200-01
// 1/14/24
// P1
// Employee.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
namespace P1
{
    class Employee
    {
    /* Class Invariants
    * 1) Employee Names should be unique and not empty
    * 2) The empWorkHours should not be negative
    * 3) The _empHasSkills should not be null
    * Implementation Invariant
    * 1) _empHasSkills requires an employee to have a skill 0 or 1 times (no
    duplicates)
    * 2) the skill level should be non-negative
    * Interface Invariant
    * 1) A default constructor is called when not enough info is provided for
    Employee
    * constructor. A basic name and empty dictionary is provided
    */
    private string _empName;
    private double _empWorkHours;
    private Dictionary<string, int> _empHasSkills;
    public string empName
    {
        get { return _empName; }
        private set { _empName = value; }
    }
    public double empWorkHours
    {
        get { return _empWorkHours; }
        private set { _empWorkHours = value; }
    }
    public Dictionary<string, int> empHasSkills
    {
        get { return _empHasSkills; }
        private set { _empHasSkills = value; }
    }
    public Employee()
    {
        empName = "Basic Employee Name";
        empWorkHours = 0.0;
        empHasSkills = new Dictionary<string, int>();
    }
    public Employee(string eName, double eWorkHours, Dictionary<string, int> eHasSkills)
    {
        empName = eName;
        empWorkHours = eWorkHours;
        empHasSkills = eHasSkills;
    }
    // If Skill of Employee > Task, Then skill estimate = taskTime / 2
    // If Skill of Employee = Task, Then skill estimate = taskTime
    // If Skill of Employee < Task , or Employee does not have skill, Then
    skill estimate = taskTime * 2
    /*Pre-Conditions:
    * 1) task is not empty and valid
    * 2) taskNeedSkills is not empty and valid
    *Post-Conditons:
    * 1) The skill difference will either be 1/2* base time, 1* base time, or
    2* base time
    * 2) The avgEstimate is total skill differences / num of skills
    * 3) A positive estimate is always returned
    */
    public double estimateTaskTime(Task task)
    {
        double avgEstimate = 0;
        foreach (var taskSkills in task.taskNeedSkills){
            string skillName = taskSkills.Key;
            int tSLevel = taskSkills.Value;

            if(empHasSkills.ContainsKey(skillName))
            {
                int empSkillLevel = empHasSkills[skillName];
                double skillDiff = 0;
                if (empSkillLevel > tSLevel)
                {
                    skillDiff = task.taskBaseTimeHours / 2;
                }
                else if (empSkillLevel < tSLevel)
                {
                    skillDiff = task.taskBaseTimeHours * 2;
                }
                else
                {
                    skillDiff = task.taskBaseTimeHours;
                }
                avgEstimate += skillDiff;
            }else{
                double skillDiff = 2 * task.taskBaseTimeHours;
                avgEstimate += skillDiff;
            }
        }
        return avgEstimate /task.taskNeedSkills.Count;
    }

    // If skill of employee < task, or employee does not have skill
    // Then skillGain = 2
    // If skill of employee >= task
    // Then skillGain = 1
    /*Pre-Conditions:
    * 1) task is not empty and valid
    * 2) task isDone == true
    * 3) taskNeedSkills is not empty and valid
    * 4) empHasSkills is not empty and valid
    *Post-Conditons:
    * 1) Task isDone will be true
    * 2) Employee skills required for task will increase by 1 or 2
    * 3) The skillGain increases by 1 if employee skill level is >= task skill
    level
    * The skillGain increases by 2 if employee does not have task skill
    * or emp skill lvl < task skill lvl
    */
    public void completeTask(Task task)
    {
        if (!task.isDone)
        {
            double estimatedTime = estimateTaskTime(task);
            task.setAsDone();
            empWorkHours += estimatedTime;
            foreach (var taskSkills in task.taskNeedSkills)
            {
                string skillName = taskSkills.Key;
                int taskDifficulty = taskSkills.Value;
                if (empHasSkills.ContainsKey(skillName))
                {
                    double empSkillLevel = empHasSkills[skillName];
                    int skillGain = 0;
                    if(empSkillLevel >= taskDifficulty)
                    {
                        skillGain = 1;
                    }
                    else
                    {
                        skillGain = 2;
                    }
                    empHasSkills[skillName] += skillGain;
                }else{
                    int skillGain = 2;
                    empHasSkills[skillName] += skillGain;
                }
            }
        }
    }
    }
}

File: Project 2/Employee.cpp
// Garland Lau
// CPSC 3200-01
// 1/30/24
// P2
// Employee.cpp

/* Class Invariants
 * - Employee Names should be unique and not empty
 * - empWorkHours should not be negative
 * - empHasSkills should not be null
 * - empHasSkills requires an employee to have a skill 0 or 1 times (no duplicates)
 * - the skill level should be non-negative
*/

#include "Employee.h"
#include "Task.h"
#include <map>
#include <unordered_map>
#include <iostream>

using namespace std;

Employee::Employee(){
    name = "No Name";
    totalWorkHours = 0;
    empSkills = {};
}

Employee::Employee(string newName, int newTotalHrs, unordered_map <string, int> newEmpSkills){
    name = newName;
    totalWorkHours = newTotalHrs;
    empSkills = newEmpSkills;
}

Employee::~Employee(){
    name = "";
    totalWorkHours = 0;
    empSkills = {};
}

Employee::Employee(const Employee& other){
    name = other.name;
    totalWorkHours = other.totalWorkHours;
    empSkills = other.empSkills;
}

Employee &Employee::operator=(const Employee& other){
    if(this == &other){
        return *this;
    }
    name = other.name;
    totalWorkHours = other.totalWorkHours;
    empSkills = other.empSkills;
    return *this;
}

/* Pre-conditions: task is valid. taskSkills is not empty
 * Post-Conditions: returns answer >= 1. Numbers will always round up (o.5 -> 1)
 */

int Employee::estimateTaskTime(const Task& task) {
    int totalSkillDiff = 0;
    int taskTime = task.getBaseTimeHours();
    int answer;

    for(const auto& entry : task.getTaskSkills()) {
        string skill_name = entry.first;
        int task_sk_level = entry.second;

        if (empSkills.find(skill_name) != empSkills.end()) {
            int emp_sk_level = get_emp_sk_lvl(skill_name);
            int tempSkillDiff = 0;
            tempSkillDiff = emp_sk_level - task_sk_level;
            totalSkillDiff += tempSkillDiff;
        }
    }

    // Task time will always be rounded up
    if(totalSkillDiff > 0) {
        if(taskTime % 2 == 1) {
            taskTime += 1;
        }
        answer = taskTime / 2;
    }else if(totalSkillDiff < 0) {
        answer = taskTime * 2;
    }else {
        answer = taskTime;
    }
    return answer;
}

/* Pre-Conditions: task is valid. taskSkills is not empty
 *  -Employee is valid. empSkills is not empty
 * Post-Conditions: task will be done.
 *  -empSkills will increase. If skill from task not in empSkills,
 *      taskSkill will be inserted into empSKills
 *  -skill level increases based on difference between task and employee
 *      levels. If employee is already proficent, increases by 1.
 *      Otherwise increase by half the difference between task and employee levels
 */

void Employee::completeTask(Task task) {
    int addHours = estimateTaskTime(task);
    totalWorkHours += addHours;
    task.setTaskAsDone();

    for(const auto& entry: task.getTaskSkills()){
        string skill_name = entry.first;
        int task_sk_lvl = entry.second;

        int skill_inc;
        if(empSkills.find(skill_name) != empSkills.end()){
            int emp_sk_lvl = get_emp_sk_lvl(skill_name);
            int skillDiff = emp_sk_lvl - task_sk_lvl;

            if(skillDiff >= 0){
                skill_inc = 1;
            }else{
                if(skillDiff % 2 == 1){
                    skillDiff += 1;
                }
                skillDiff = skillDiff / 2;
            }
            skill_inc += emp_sk_lvl;
            empSkills[skill_name] = skill_inc;
        }else{
            int skillDiff = task_sk_lvl;
            skill_inc;
            if (skillDiff % 2 == 1){
                skillDiff += 1;
            }
            skill_inc = skillDiff / 2;
            empSkills[skill_name] = skill_inc;
        }
    }
}